packages/api-client-core/src/GadgetConnection.ts

```diff
@@ -350,4 +321,50 @@ export class GadgetConnection {
 
     return headers;
   }
+
+  private async waitForOpenedConnection(
+    attempts: number,
+    timeoutMs: number,
+    overrides: GadgetSubscriptionClientOptions
+  ): Promise<SubscriptionClient> {
+    let subscriptionClient = this.newSubscriptionClient(overrides);
+    let unsubscribes: Function[] = []; // eslint-disable-line @typescript-eslint/ban-types
+
+    const clearListeners = () => {
+      unsubscribes.forEach((fn) => fn());
+      unsubscribes = [];
+    };
+
+    return await new Promise<SubscriptionClient>((resolve, reject) => {
+      const wrappedReject = (err: any) => {
+        if (err && err.code == CloseCode.ConnectionAcknowledgementTimeout && attempts > 0) {
+          attempts -= 1;
+          subscriptionClient = this.newSubscriptionClient(overrides);
```

Comment:
We should probably ` subscriptionClient.dispose` the old client, eh?

Reply:
ohh good call, I'll add that